src/axom/inlet/tests/inlet_object.cpp

```diff
@@ -0,0 +1,273 @@
+// Copyright (c) 2017-2020, Lawrence Livermore National Security, LLC and
+// other Axom Project Developers. See the top-level COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+#include "gtest/gtest.h"
+
+#include <string>
+#include <vector>
+#include <unordered_map>
+
+#include <iostream>
+
+#include "axom/sidre.hpp"
+
+#include "axom/inlet/LuaReader.hpp"
+#include "axom/inlet/Inlet.hpp"
+
+using axom::inlet::Inlet;
+using axom::inlet::LuaReader;
+using axom::sidre::DataStore;
+
+std::shared_ptr<Inlet> createBasicInlet(DataStore* ds,
+                                        const std::string& luaString,
+                                        bool enableDocs = true)
+{
+  auto lr = std::make_shared<LuaReader>();
+  lr->parseString(luaString);
+
+  return std::make_shared<Inlet>(lr, ds->getRoot(), enableDocs);
+}
+
+struct Foo
+{
+  bool bar;
+  bool baz;
+};
+
+void from_inlet(axom::inlet::Table& base, Foo& f)
+{
+  f.bar = base["bar"];
+  f.baz = base["baz"];
+}
+
+TEST(inlet_object, simple_struct_by_ref)
+{
+  std::string testString = "foo = { bar = true; baz = false }";
+  DataStore ds;
+  auto inlet = createBasicInlet(&ds, testString);
+
+  // Define schema
+  std::shared_ptr<axom::inlet::Field> currField;
+
+  // Check for existing fields
+  currField = inlet->addBool("foo/bar", "bar's description");
+  EXPECT_TRUE(currField);
+
+  currField = inlet->addBool("foo/baz", "baz's description");
+  EXPECT_TRUE(currField);
+
+  Foo foo;
+  bool found = inlet->get("foo", foo);
+
+  EXPECT_TRUE(found);
+  EXPECT_TRUE(foo.bar);
+  EXPECT_FALSE(foo.baz);
+}
+
+template <>
+Foo from_inlet<Foo>(axom::inlet::Table& base)
```

Comment:
So, it would be up to the user to choose which version to define ?

Reply:
Yes, the template specialization version that returns a T is intended for move-only/non-default-constructible types, though other types could use that version as well.  It is possible to change the by-reference get (soon to be renamed to `get_to`) to deduce at compile time which version was defined for a particular type since it can be implemented in terms of either, so I may look into implementing that.

Reply:
I have added in some more deduction logic that makes these specializations/overloads significantly more flexible and no longer defers existence checks until link time - they can be checked as part of template instantiation.